File: Data Structures & Algorithms/subsets/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) {
        let res = [];
        let subset = [];
        this.dfs(0, nums, subset, res);
        return res;
    }

    dfs(i, nums, subset, res) {
        if (i == nums.length) {
            res.push([...subset]);
            return;
        }

        subset.push(nums[i]);
        this.dfs(i + 1, nums, subset, res);

        subset.pop();
        this.dfs(i + 1, nums, subset, res);

        }

    
}

File: Data Structures & Algorithms/subsets-ii/submission-1.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        subset = []

        def dfs(i, subset, res):
            if (i == len(nums)) and (subset not in res):
                res.append(subset.copy())
                return
            elif (i == len(nums)) and (subset in res):
                return

            subset.append(nums[i])
            dfs(i + 1, subset, res)

            subset.pop()
            dfs(i + 1, subset, res)

        dfs(0, subset, res)

        return res




        